packages/expo-cli/src/commands/build/ios/IOSBuilder.ts

```diff
@@ -144,21 +145,102 @@ See https://docs.expo.io/versions/latest/distribution/building-standalone-apps/#
           ErrorCodes.NON_INTERACTIVE,
           'Unable to proceed, see the above error message.'
         );
+      }
+
+      log(
+        chalk.bold.red(
+          'Failed to prepare all credentials. \nThe next time you build, we will automatically use the following configuration:'
+        )
+      );
+      throw e;
+    } finally {
+      const credentials = await context.ios.getAllCredentials();
+      displayProjectCredentials(experienceName, bundleIdentifier, credentials);
+    }
+  }
+
+  async _setupDistCert(
+    ctx: Context,
+    experienceName: string,
+    bundleIdentifier: string,
+    appCredentials: IosAppCredentials
+  ): Promise<void> {
+    try {
+      const nonInteractive = this.options.parent && this.options.parent.nonInteractive;
+      const distCertFromParams = await getDistCertFromParams(this.options);
+      if (distCertFromParams) {
+        await useDistCertFromParams(ctx, appCredentials, distCertFromParams);
       } else {
-        log(
-          chalk.bold.red(
-            'Failed to prepare all credentials. \nThe next time you build, we will automatically use the following configuration:'
-          )
+        await runCredentialsManager(
+          ctx,
+          new SetupIosDist({ experienceName, bundleIdentifier, nonInteractive })
         );
       }
+    } catch (e) {
+      log.error('Failed to set up Distribution Certificate');
+      throw e;
     }
+  }
 
-    const credentials = await context.ios.getAllCredentials();
-    displayProjectCredentials(experienceName, bundleIdentifier, credentials);
+  async _setupPushCert(
+    ctx: Context,
+    experienceName: string,
+    bundleIdentifier: string,
+    appCredentials: IosAppCredentials
+  ): Promise<void> {
+    try {
+      const nonInteractive = this.options.parent && this.options.parent.nonInteractive;
+      const pushKeyFromParams = await getPushKeyFromParams(this.options);
+      if (pushKeyFromParams) {
+        await usePushKeyFromParams(ctx, appCredentials, pushKeyFromParams);
+      } else {
+        await runCredentialsManager(
+          ctx,
+          new SetupIosPush({ experienceName, bundleIdentifier, nonInteractive })
+        );
+      }
+    } catch (e) {
+      log.error('Failed to set up Push Key');
```

Comment:
is there any more context that we can provide in this error?

Reply:
The more context-specific error with the `expo.fyi` link is caught and rethrown (which will then print to stderr). I catch it at this point and print out `Failed to set up Push Key` so we know which part of the credential process we failed at. The error logging would look something like this:

```
Failed to set up Push Key
Additional information needed to setup credentials in non-interactive mode.
Learn more about how to resolve this: expo.fyi/credentials-non-interactive (​https://expo.fyi/credentials-non-interactive​).
```